### PR TITLE
feat: sending location messages (WPB-1732)

### PR DESCRIPTION
**Pull request by yamilmedina**
<table>
<td>
  <a href="https://wearezeta.atlassian.net/browse/WPB-1732" title="WPB-1732" target="_blank">StoryWPB-1732</a>  [Android] Share Location - sending
  </td></table>
  <br />
 

----
#### PR Submission Checklist for internal contributors

- The **PR Title**
  - [x] conforms to the style of semantic commits messages¹ supported in Wire's Github Workflow²
  - [x] contains a reference JIRA issue number like `SQPIT-764`
  - [x] answers the question: _If merged, this PR will: ..._ ³

- The **PR Description**
  - [x] is free of optional paragraphs and you have filled the relevant parts to the best of your ability
----

# What's new in this PR?

### Issues

We didn't have support for sending location messages.

### Causes (Optional)

Not implemented.

### Solutions

Implement the sending location messages (supporting self deleting as well) 

### Dependencies (Optional)

- https://github.com/wireapp/kalium/pull/2363

Needs releases with:

- [x] GitHub link to other pull request

#### Test Coverage (Optional)

- [ ] I have added automated test to this contribution

### Attachments

**GMS**

https://github.com/wireapp/wire-android/assets/5806454/935562d6-ecb0-484a-a44f-87bc5f7819ff

**NO GMS**

https://github.com/wireapp/wire-android/assets/5806454/cd0f17bb-51c9-4db3-b8ec-b801c27eedd6

----
#### PR Post Submission Checklist for internal contributors (Optional)

 - [x] Wire's Github Workflow has automatically linked the PR to a JIRA issue
----
#### PR Post Merge Checklist for internal contributors

 - [ ] If any soft of configuration variable was introduced by this PR, it has been added to the relevant documents and the CI jobs have been updated.
----
##### References
1. https://sparkbox.com/foundry/semantic_commit_messages
1. https://github.com/wireapp/.github#usage
1. E.g. `feat(conversation-list): Sort conversations by most emojis in the title #SQPIT-764`.

### CODE REVIEW

**Comment by AndroidBob** [comment]
Build [2557](https://10.10.124.134/job/AR-build-pipeline/2557/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.6.0(8637385)(8637385)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.6.0(8637385)(8637385)-staging-compat.apk)

File: app/src/main/kotlin/com/wire/android/util/permission/CurrentLocationRequestFlow.kt [hunk 1]
```diff
@@ -29,34 +29,47 @@ import com.wire.android.util.extension.checkPermission
...
-            if (isGranted) {
-                // TODO: launch map location picker using openstreetmap? (have in mind f-droid aka. no gms)
+    val requestPermissionLauncher: ManagedActivityResultLauncher<Array<String>, Map<String, @JvmSuppressWildcards Boolean>> =
+        rememberLauncherForActivityResult(ActivityResultContracts.RequestMultiplePermissions()) { permissions ->
+            val allPermissionGranted = permissions.all { it.value }
+            if (allPermissionGranted) {
+                onPermissionAllowed()
             } else {
                 onPermissionDenied()
             }
```
Comment (ohassine) [inline app/src/main/kotlin/com/wire/android/util/permission/CurrentLocationRequestFlow.kt:44]: I think here we don't need to show the settings dialog every time the user deny the location permission, that dialog should be shown when it's permanently denied only. 
We can check that using `shouldShowRequestPermissionRationale()`

Comment (yamilmedina) [reply app/src/main/kotlin/com/wire/android/util/permission/CurrentLocationRequestFlow.kt:44]: Thanks, good point. 
According to [designs](https://www.figma.com/file/UgKehvwIbiIiQ2CCEdsVuf/Android?node-id=40141%3A75324&mode=dev), we need to show this dialog if the user declined previously (basically we are suggesting him to go to "app settings" with redirection). Also, this is the same behavior for other attachments like Recording Audio.

Comment (yamilmedina) [reply app/src/main/kotlin/com/wire/android/util/permission/CurrentLocationRequestFlow.kt:44]: Good idea about https://wearezeta.atlassian.net/browse/WPB-6154
I extracted a component, so we can reuse it at.

app/src/main/kotlin/com/wire/android/util/permission/PermissionsDeniedRequestDialog.kt

Comment (ohassine) [reply app/src/main/kotlin/com/wire/android/util/permission/CurrentLocationRequestFlow.kt:44]: I think it's not up to date
https://www.figma.com/file/UgKehvwIbiIiQ2CCEdsVuf/Android?type=design&node-id=10548-105517&mode=dev
Also it's recommended to do it this way
https://developer.android.com/training/permissions/requesting#handle-denial

Comment (yamilmedina) [reply app/src/main/kotlin/com/wire/android/util/permission/CurrentLocationRequestFlow.kt:44]: I'll leave it as is, since it's up-to-date for the sharing location MVP and congruent with the recording audio approach of handling (the most recent attachment implementation we have). Anyway, we will take a look in the design review for more comments from the design team in case changes are needed =)

Also, since a new ticket was created. I think the recommended way to handle the permissions and the figma file linked there is in the scope of this other ticket https://wearezeta.atlassian.net/browse/WPB-6154

**Comment by AndroidBob** [comment]
Build [2578](https://10.10.124.134/job/AR-build-pipeline/2578/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.6.0(8643852)(8643852)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.6.0(8643852)(8643852)-staging-compat.apk)

**Comment by AndroidBob** [comment]
Build [2581](https://10.10.124.134/job/AR-build-pipeline/2581/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.6.0(8644399)(8644399)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.6.0(8644399)(8644399)-staging-compat.apk)

**Review by borichellow — APPROVED** [review]

File: app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/location/LocationPickerViewModel.kt [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+import android.content.Context
+import android.location.Geocoder
+import androidx.compose.runtime.getValue
+import androidx.compose.runtime.mutableStateOf
+import androidx.compose.runtime.setValue
+import androidx.lifecycle.ViewModel
+import androidx.lifecycle.viewModelScope
+import com.google.android.gms.location.LocationServices
+import com.google.android.gms.location.Priority.PRIORITY_HIGH_ACCURACY
+import com.google.android.gms.tasks.CancellationTokenSource
```
Comment (vitorhugods) [inline app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/location/LocationPickerViewModel.kt:33]: A heads-up: this won't work on devices without Google Play Services

**Review by vitorhugods — COMMENTED** [review]
It looks good to me.

Just curious about what are the plans regarding F-Droid and devices without Google Play Services.

Maybe abstract the violating code behind an interface, so we can more easily inject/replace it later?

If the plan is to do this later, I'm fine with it too. Just raising concern here, really.

File: app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/location/LocationPickerViewModel.kt (hunk 2 above)
Comment (yamilmedina) [reply app/src/main/kotlin/com/wire/android/ui/home/messagecomposer/location/LocationPickerViewModel.kt:33]: Good catch, I'm going to add a check for playservices, so we use a fallback (from which we are only going to be able to get `lat`, `lon`), but at least that's something :D 

Regarding F-Droid and other devices you can take a look at the feature [breakdown](https://wearezeta.atlassian.net/wiki/spaces/~751534962/pages/939196449/20231114+-+Location+feature+breakdown), the [epic](https://wearezeta.atlassian.net/browse/WPB-5169) and the prod [feature page ](https://wearezeta.atlassian.net/wiki/spaces/PROD/pages/946143233/Share+Location) to see how we planned for it.

The TL;DR; of it is, problems for the future, when the time comes, we have tickets for addressing those.

EDIT: Attached a video, from a non-gms device, the only alternative we have, but at least is something

**Comment by AndroidBob** [comment]
Build [2593](https://10.10.124.134/job/AR-build-pipeline/2593/) **failed**.

**Comment by AndroidBob** [comment]
Build [2597](https://10.10.124.134/job/AR-build-pipeline/2597/) **succeeded**.

The build produced the following APK's:
- [com.wire.android-v4.6.0(8645923)(8645923)-staging-compat.apk](https://z-lohika.s3-eu-west-1.amazonaws.com/megazord/android/reloaded/staging/compat/com.wire.android-v4.6.0(8645923)(8645923)-staging-compat.apk)

**Review by MohamadJaara — APPROVED** [review]
